### ISSUE TITLE
Active learning step: ValueError: operands could not be broadcast together with shapes (10,) (9,)

### ISSUE THREAD

**Issue by yikezhu1** [YttriLab/A-SOID#32]
Hi there! While we are trying to do the active learning step, we keep encountering the same value error, the full error is below:

ValueError: operands could not be broadcast together with shapes (10,) (9,)
Traceback:
File "C:\Users\kkw11\.conda\envs\asoid-test\lib\site-packages\streamlit\scriptrunner\script_runner.py", line 557, in _run_script
    exec(code, module.__dict__)
File "C:\Users\kkw11\.conda\envs\asoid-test\lib\site-packages\asoid\app.py", line 332, in <module>
    main()
File "C:\Users\kkw11\.conda\envs\asoid-test\lib\site-packages\asoid\app.py", line 328, in main
    application_function()
File "C:\Users\kkw11\.conda\envs\asoid-test\lib\site-packages\asoid\apps\C_auto_active_learning.py", line 128, in main
    rf_classifier.main()
File "C:\Users\kkw11\.conda\envs\asoid-test\lib\site-packages\asoid\utils\auto_active_learning.py", line 549, in main
    self.base_classification()
File "C:\Users\kkw11\.conda\envs\asoid-test\lib\site-packages\asoid\utils\auto_active_learning.py", line 326, in base_classification
    self.show_subsampled_performance()
File "C:\Users\kkw11\.conda\envs\asoid-test\lib\site-packages\asoid\utils\auto_active_learning.py", line 281, in show_subsampled_performance
    mean_scores2beat = np.mean(np.mean(self.all_f1_scores, axis=0), axis=0)
File "<__array_function__ internals>", line 180, in mean
File "C:\Users\kkw11\.conda\envs\asoid-test\lib\site-packages\numpy\core\fromnumeric.py", line 3432, in mean
    return _methods._mean(a, axis=axis, dtype=dtype,
File "C:\Users\kkw11\.conda\envs\asoid-test\lib\site-packages\numpy\core\_methods.py", line 180, in _mean
    ret = umr_sum(arr, axis, dtype, out, keepdims, where=where)

We tried to keep the video lengths of the two annotation CSV files coherent, and no overlapping behaviors at the same time point. Would you please provide us some instruction on how to proceed? Thank you so much!

**Comment by JensBlack** [comment]
Thanks for using A-SOiD!

It looks like you do not have the same amount of label and pose files imported. You can verify if you imported the same number of files for both during data import by looking at the corresponding config.ini parameters.

Here is an example of one of my own projects with 2 each:

```
[Data]
DATA_INPUT_FILES = 16726_SIT_1.analysis.h5, 16726_SIT_2.analysis.h5

LABEL_INPUT_FILES = 16726_T1_Resident.csv, 16726_T2_Resident.csv

ROOT_PATH = None
```

If that is not the issue, I will require additional details to get to the root of your issue.
Please post the content of your config.ini file here and the first few rows of an example label and pose file.

**Comment by JensBlack** [comment]
I found a similar bug #33. It would be great if you could report your issue in the same way.

**Comment by karimabdelaal** [comment]
Hello,

@yikezhu1 is a trainee of mine that has been helping me get ASOID up and running for our projects. This issue has been resolved, however we are not exactly sure why.

We re-tracked our animal dataset using DLC and reset the ASOID app and the issue has not arose since. An interesting thing we noticed is that if we reuse a path, even if we make a new config file and/or update the previous file, we always get this error. Resetting the app usually helps.

I hope this is useful, I am happy to elaborate and clarify any other details. My apologies for the delay on this response.

**Comment by yikezhu1** [comment]
[like]  Yike Zhu reacted to your message:
________________________________

**Comment by JensBlack** [comment]
> Hello,
> 
> @yikezhu1 is a trainee of mine that has been helping me get ASOID up and running for our projects. This issue has been resolved, however we are not exactly sure why.
> 
> We re-tracked our animal dataset using DLC and reset the ASOID app and the issue has not arose since. An interesting thing we noticed is that if we reuse a path, even if we make a new config file and/or update the previous file, we always get this error. Resetting the app usually helps.
> 
> I hope this is useful, I am happy to elaborate and clarify any other details. My apologies for the delay on this response.

Great that you found a solution, however, it would be great if you could elaborate on the "reset" of the app and post the content of both config files.

This way any user experiencing the same issue has the chance to resolve it and I can solve it for the next update!

Thanks!

**Comment by karimabdelaal** [comment]
Hello. I will do my best to elaborate on the changes we made that seemed to resolve the issues. One other change I forgot to mention is that our BORIS annotation files sometimes had 2 behaviors co-occurring at once. We did fix this so that for each time frame, only a single behavior was labeled. This originally did not work, however, after completely closing the app and closing the terminal code, we restarted everything and it ran smoothly. 

In terms of the path issue, this was particularly occurring on our windows system. What happened is that we started an ASOID project, uploaded the data and extracted the frames fine. We would then run into the stated error above during the Active Learning step. In attempts to get things working, we were relabeling things on BORIS and re-tracking DLC. We tried to upload new data into the config file and extract features but were running into errors (can't remember the error off the top of my head). We then deleted the entire output folder that the config, data, and feats_targets files were saved in and tried again but would still run into the same error. The only fixes we could find were closing the ASOID app and reopening it, or using a different path.

The newest issue that I am running into that is resolved most of the time by closing and reopening the app is below (I am happy to open a new issue for this if useful). This is specifically happening when we try to do active learning after the discover step. We train the model, look at the unsupervised classification and add a new class based on the clustering results of our 'other' group. It saves a new config file into a new folder and when we try to run active learning on that data, I get this error. Two things I have noted is that features need to be re-extracted and after extracting features, if I get this error in the active learning step, I reset the app and it usually works. I haven't fully pinned down when exactly the issue is occurring and when the reset is needed to help fix it, but so far that has been our go-to fix and it has been working well. The config files for the original dataset and the new config file that is produced after adding the new classes are attached as screenshots below.

```
ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (10,) + inhomogeneous part.
File "/home/ka234/anaconda3/envs/asoid-test/lib/python3.8/site-packages/streamlit/scriptrunner/script_runner.py", line 557, in _run_script
    exec(code, module.__dict__)
File "/home/ka234/anaconda3/envs/asoid-test/lib/python3.8/site-packages/asoid/app.py", line 332, in <module>
    main()
File "/home/ka234/anaconda3/envs/asoid-test/lib/python3.8/site-packages/asoid/app.py", line 328, in main
    application_function()
File "/home/ka234/anaconda3/envs/asoid-test/lib/python3.8/site-packages/asoid/apps/C_auto_active_learning.py", line 128, in main
    rf_classifier.main()
File "/home/ka234/anaconda3/envs/asoid-test/lib/python3.8/site-packages/asoid/utils/auto_active_learning.py", line 549, in main
    self.base_classification()
File "/home/ka234/anaconda3/envs/asoid-test/lib/python3.8/site-packages/asoid/utils/auto_active_learning.py", line 326, in base_classification
    self.show_subsampled_performance()
File "/home/ka234/anaconda3/envs/asoid-test/lib/python3.8/site-packages/asoid/utils/auto_active_learning.py", line 281, in show_subsampled_performance
    mean_scores2beat = np.mean(np.mean(self.all_f1_scores, axis=0), axis=0)
File "<__array_function__ internals>", line 200, in mean
File "/home/ka234/anaconda3/envs/asoid-test/lib/python3.8/site-packages/numpy/core/fromnumeric.py", line 3464, in mean
    return _methods._mean(a, axis=axis, dtype=dtype,
File "/home/ka234/anaconda3/envs/asoid-test/lib/python3.8/site-packages/numpy/core/_methods.py", line 165, in _mean
    arr = asanyarray(a)
```

![Screenshot_2023-06-09_13-56-33](https://github.com/YttriLab/A-SOID/assets/135879190/90683812-5920-4a2c-bba7-06770a84d589)
![Screenshot_2023-06-09_13-57-44](https://github.com/YttriLab/A-SOID/assets/135879190/f9f10d95-7e74-4b9b-ada9-d8dac0ada8a6)

**Comment by yikezhu1** [comment]
[like]  Yike Zhu reacted to your message:
________________________________

**Comment by JensBlack** [comment]
> One other change I forgot to mention is that our BORIS annotation files sometimes had 2 behaviors co-occurring at once.

Yes, A-SOiD assumes that behavioral states are exclusive. Good Catch!

> The only fixes we could find were closing the ASOID app and reopening it or using a different path.

Streamlit apps have a caching system that sometimes believes it's smarter than the developer and user, so it might be that some of the data that you were trying to reload were saved in the cache and not updated. Restarting the app is the only non-hacky way to solve this for now. I will look into this particular issue more.

> Two things I have noted is that features need to be re-extracted and after extracting features

The altered annotations (including the cluster) are exported to a new project to reduce redundancy. Unfortunately, you are not the only one that experiences some issues with this step. We are releasing a substantial update to the workflow that should fix most issues and comes with additional functionalities with the new update.

**It would be great if we could split both issues for clarity. Can you please open a new issue? Thanks!**
